File: routes/reports.js
const express = require('express');
const router = express.Router();
const db = require('../database');

router.get('/reports', async (req, res) => {
  try {
    const result = await db.query('SELECT * FROM reports ORDER BY created_at DESC');

    const reports = result.rows.map(row => ({
      id: row.id,
      name: row.name,
      description: row.description,
      filters: row.filters,
      createdAt: row.created_at,
      lastModified: row.last_modified
    }));

    res.json(reports);
  } catch (error) {
    console.error('Error reading reports:', error);
    res.status(500).json({ error: 'Failed to read reports' });
  }
});

router.post('/reports', async (req, res) => {
  try {
    const { report } = req.body;

    await db.query(
      `INSERT INTO reports (id, name, description, filters, created_at, last_modified)
       VALUES ($1, $2, $3, $4, $5, $6)
       ON CONFLICT (id) DO UPDATE SET
         name = EXCLUDED.name,
         description = EXCLUDED.description,
         filters = EXCLUDED.filters,
         last_modified = EXCLUDED.last_modified`,
      [
        report.id,
        report.name,
        report.description || null,
        JSON.stringify(report.filters || {}),
        report.createdAt || new Date().toISOString(),
        report.lastModified || new Date().toISOString()
      ]
    );

    res.json({ success: true, reportId: report.id });
  } catch (error) {
    console.error('Error saving report:', error);
    res.status(500).json({ error: 'Failed to save report' });
  }
});

router.delete('/reports/:reportId', async (req, res) => {
  try {
    const { reportId } = req.params;

    await db.query('DELETE FROM reports WHERE id = $1', [reportId]);

    res.json({ success: true });
  } catch (error) {
    console.error('Error deleting report:', error);
    res.status(500).json({ error: 'Failed to delete report' });
  }
});

router.post('/reports/:reportId/data', async (req, res) => {
  try {
    res.json({ success: true });
  } catch (error) {
    console.error('Error saving report data:', error);
    res.status(500).json({ error: 'Failed to save report data' });
  }
});

router.get('/reports/:reportId/data', async (req, res) => {
  try {
    return res.status(404).json({ error: 'Report data not found' });
  } catch (error) {
    console.error('Error reading report data:', error);
    res.status(500).json({ error: 'Failed to read report data' });
  }
});

module.exports = router;

File: backend/app/routes/reports.py
from datetime import UTC, datetime

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy import delete, select
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.auth import get_current_household_id
from app.models.report import Report
from app.schemas.report import ReportOut, ReportSaveRequest

router = APIRouter(prefix="/api", tags=["reports"])


@router.get("/reports")
async def get_reports(
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Report)
        .where(Report.household_id == household_id)
        .order_by(Report.created_at.desc())
    )
    return [ReportOut.from_orm_model(r).model_dump() for r in result.scalars().all()]


@router.post("/reports")
async def save_report(
    body: ReportSaveRequest,
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    r = body.report
    now = datetime.now(UTC).replace(tzinfo=None)
    created = datetime.fromisoformat(r.createdAt) if r.createdAt else now
    modified = datetime.fromisoformat(r.lastModified) if r.lastModified else now
    stmt = (
        insert(Report)
        .values(
            id=r.id,
            name=r.name,
            description=r.description,
            household_id=household_id,
            filters=r.filters or {},
            created_at=created,
            last_modified=modified,
        )
        .on_conflict_do_update(
            index_elements=[Report.id],
            set_={
                "name": r.name,
                "description": r.description,
                "filters": r.filters or {},
                "last_modified": modified,
            },
        )
    )
    await db.execute(stmt)
    await db.commit()

    return {"success": True, "reportId": r.id}


@router.delete("/reports/{report_id}")
async def delete_report(
    report_id: str,
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(
        delete(Report).where(
            Report.id == report_id, Report.household_id == household_id
        )
    )
    await db.commit()
    return {"success": True}


@router.post("/reports/{report_id}/data")
async def save_report_data(report_id: str):
    return {"success": True}


@router.get("/reports/{report_id}/data")
async def get_report_data(report_id: str):
    return JSONResponse(status_code=404, content={"error": "Report data not found"})
